### ISSUE TITLE
glGetUniformLocation (1, b’Texture\x00’) 'invalid operation'

### ISSUE THREAD

**Issue by Willy-JL** [pyimgui/pyimgui#309]
First of all, I'm very excited for the close release of 2.0, and I'm sorry to be submitting this issue at this time. I'm fairly certain it's a problem with the OS and not pyimgui, but I couldn't find any concrete results online...

```
Traceback (most recent call last):
  ...
  File "imgui/integrations/glfw.py", line 12, in __init__
  File "imgui/integrations/opengl.py", line 60, in __init__
  File "imgui/integrations/base.py", line 17, in __init__
  File "imgui/integrations/opengl.py", line 108, in _create_device_objects
  File "src/latebind.pyx", line 51, in OpenGL_accelerate.latebind.Curry.__call__
  File "OpenGL/GL/VERSION/GL_2_0.py", line 430, in glGetUniformLocation
  File "OpenGL/platform/baseplatform.py", line 415, in __call__
  File "src/errorchecker.pyx", line 58, in OpenGL_accelerate.errorchecker._ErrorChecker.glCheckError
OpenGL.error.GLError: GLError(
	err = 1282,
	description = b'invalid operation',
	baseOperation = glGetUniformLocation,
	cArguments = (1, b'Texture\x00'),
	result = -1
)
```
From what I can tell the issue comes up [here](https://github.com/pyimgui/pyimgui/blob/dev/version-2.0/imgui/integrations/opengl.py#L108), even though the traceback doesn't state it precisely.

I haven't had this issue on any of my systems, and most of my users had no issues with opengl, glfw and imgui. This issue was reported by only 2 users, and both on a Debian based distribution of linux (one on Devuan and one on either Pop!_OS or base Debian). That's why my best guess is that this error is due to some caveat with debian based packages, but I could very well be wrong. From what I recall the first report came up maybe 6 months ago, while the second was 2 months ago. Both users have been completely unable to get a window to show up with my program, even through my regular updates (and I always use the dev/version-2.0 branch). I package it with cx_Freeze, but the issue happens also when running from source.

In case this matters, I know that the user on Devuan is using an NVidia gpu with nouveau drivers, I don't know about the other report.

I had poked at it a few times before with no results, and now I tried to look harder and still managed nothing. Any help, even just tiny hints, is very welcome

**Comment by Willy-JL** [comment]
From what i found online, the `#version 330` sets which GLSL version (`GL_SHADING_LANGUAGE_VERSION`) to use, and usually the opengl version goes hand in hand with that, for example on my system `glinfo` output:
```
GL_VERSION: 4.6.0 NVIDIA 520.56.06
GL_RENDERER: NVIDIA GeForce RTX 2060/PCIe/SSE2
GL_VENDOR: NVIDIA Corporation
GL_EXTENSIONS: ...
GL_SHADING_LANGUAGE_VERSION = 4.60 NVIDIA
GLU_VERSION: 1.3
GLU_EXTENSIONS: ...
GLUT_API_VERSION: 4
GLUT_XLIB_IMPLEMENTATION: 13
```
and you can see that opengl version 4.6.0 goes hand in hand with the glsl version 4.60. i thought that the integration used a window hint to specify the gl version to use, but [`impl_glfw_init`](https://github.com/pyimgui/pyimgui/blob/480b4de8ab82adaa2cd310100330b41f40274bcc/doc/examples/integrations_glfw3.py#L55-L81) ~~only does that on macos...~~ it used to do that only on macos when i originally copied it into my project, but i see it was changed to always set the window hints...

i will update my `impl_glfw_init`, ask the two user to try that and also send the output of `glinfo` for more context. will report back soon, and thanks.

**Comment by Willy-JL** [comment]
Looks like that was it! Ran into some more struggles due to typical debian packages being held back, but after that its running just fine! Thanks again.

For now only one of the two users has reported back, but I'm confident this was it, if not I'll comment again

**Comment by Willy-JL** [comment]
Welp, looks like I have been cursed. While that fixed it for one linux user, it created another issue for a windows user...

```
Traceback (most recent call last):
  ...
  File "imgui\integrations\glfw.py", line 12, in __init__
  File "imgui\integrations\opengl.py", line 60, in __init__
  File "imgui\integrations\base.py", line 17, in __init__
  File "imgui\integrations\opengl.py", line 84, in _create_device_objects
  File "src\latebind.pyx", line 39, in OpenGL_accelerate.latebind.LateBind.__call__
  File "src\wrapper.pyx", line 318, in OpenGL_accelerate.wrapper.Wrapper.__call__
  File "src\wrapper.pyx", line 311, in OpenGL_accelerate.wrapper.Wrapper.__call__
  File "OpenGL\platform\baseplatform.py", line 415, in __call__
  File "src\errorchecker.pyx", line 58, in OpenGL_accelerate.errorchecker._ErrorChecker.glCheckError
OpenGL.error.GLError: GLError(
    err = 1282,
    description = b'invalid operation',
    baseOperation = glGetIntegerv,
    pyArgs = (
        GL_TEXTURE_BINDING_2D,
        <object object at 0x0000020BD9A54E20>,
    ),
    cArgs = (
        GL_TEXTURE_BINDING_2D,
        array([0], dtype=int32),
    ),
    cArguments = (
        GL_TEXTURE_BINDING_2D,
        array([0], dtype=int32),
    )
)
```

The previous version of my program, where the context version hints were not being set, worked fine for them and opengl was using version 4.6. But apparently setting it to 3.3 breaks it for them... Like I can understand the previous error where the implementation setup was designed around OGL 3.3, and for one particular user on a newer 4.2 version it caused issues, but here the opposite is happening where setting it to the correct version breaks it while the newer version does not...?

I was tempted to make a try except block and have it attempt 3.3 first and if that fails let it use the default version, but the problem is both errors dont happen at context creation, but much later, so I'd have to stick the whole GlfwRenderer() setup in the try block, and that would require me to enable OpenGL.ERROR_CHECKING which is a huge performance hog, and cant be disabled after the first opengl import...

**Comment by KinoxKlark** [comment]
Yes indeed, using the hint to ensure correct version (3.3 for GLSL compatibility) seems to be the proper way to fix that.

But I am not sure why it would cause some incompatibilities with 4.6 users. `glGetIntegerv` should work the same in 3.3 and 4.x, as far as I know. (It is however reported to be very slow and may be useless in the present case, thus it could be avoided). 

There have been some changes to the integration in #296, which I just merged. It may be useful to test that this doesn't fix it miraculously or breaks it even more.

**Comment by Willy-JL** [comment]
Thanks for the insight. I now have received another report of the same issue. This user has the same exact situation as the other, was working fine, and my update with the hints showed this issue, same lines in the traceback. They mentioned updating their nvidia graphics drivers yesterday, not sure if that was before or after updating my app, so I’m not sure if that driver opened the app correctly without the hints. Either way this makes me even more sure that this is some weird nvidia shenanigans, because when is nvidia ever not the problem am I right? 🙄 

I’ll try a build with the updated integration and ask the affected user to report back.

**Comment by Willy-JL** [comment]
So while adventuring down the rabbit hole that is page 2 of google i found some mentions of msi afterburner and riva tuner statistics server overlays injecting into opengl applications and causing crashes on nvidia drivers. So i asked both users and indeed, disabling rtss fixed the crash. Also updating to the latest beta of afterburner and rtss fixed the crash while keeping them enabled. So maybe a note could be added somewhere? Either way im happy this is working now, and that its not a problem with my or this great project's code.
<br>
**TLDR of solutions:**
- **make sure you enable the hints for the correct opengl version when creating your window**
- **some crashes might be caused by msi afterburner and rtss, try updating or disabling them**
